plugins/powerd/fu-powerd-plugin.c

```diff
@@ -9,6 +9,8 @@
 
 #include "fu-powerd-plugin.h"
 
+#define POWERD_SUSPEND_ANNOUNCED_PATH  "/run/power_manager/power/suspend_announced"
```

Comment:
Rather than being "reactive" I'd much rather this was "proactive" -- i.e. rather than starting the update and then aborting after the file has been downloaded, decompressed, parsed and the device put into update mode -- I'd much rather the daemon didn't even try to deploy the update in the first place. If you create a GFileMonitor pointing to this file, you can connect to the `changed` signal and then add/remove the `FU_CONTEXT_FLAG_SYSTEM_INHIBIT` flag from the per-system context instead.

Reply:
Hi hughsie,

I'm not sure if it's clear but this new code won't abort anything - it just hangs inside the loop in `fu_powerd_plugin_block_suspend ` until that file at `POWERD_SUSPEND_ANNOUNCED_PATH ` goes away.

But to your bigger point about registering a file watcher -

I'm concerned about this race condition:
- fwupd: starts up
- fwupd: GFileMonitor begins watching for suspend_announced
- fwupd: begins an update
- powerd: starts a suspend, creating suspend_announced
- fwupd: fu-powerd-plugin creates a lockfile
- fwupd: starts uploading firmware to a device
- powerd: suspends the system
- (the system resumes)
- GFileMonitor notices suspend_announced, and the callback sets FU_CONTEXT_FLAG_SYSTEM_INHIBIT (albeit too late)

(I don't know how GFileMonitor works, so maybe my assumptions about pathological timing don't apply - let me know)

WDYT?

Reply:
> fwupd: begins an update
> powerd: starts a suspend, creating suspend_announced

It's not possible to prevent this race unless fwupd can tell the system before starting "don't suspend" -- i.e. it has to come from the initiator rather than the executor. See https://github.com/fwupd/fwupd/blob/main/plugins/logind/fu-logind-plugin.c#L71 for what we do with logind.

Anything that's coming from the executor to the initiator has to either wait for a "reply", saying "_yes, suspend"_ or "_no, busy_". Blocking in the plugin mid-update is a bad idea and means that everything depending on the mainloop running is frozen. For instance, it would be very easy to get into a live-lock condition if anything even indirectly was watching fwupd for a DBus signal before continuing with the power save action.

Reply:
Sorry for going quiet, just an update: I'm going to start looking at seeing how we can coordinate a handshake with fwupd like the way logind does as you've suggested. I'll write back here with any questions that I have (or let me know if there is a better place to ask them).

Reply:
Here is fine, thanks.